src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+        return -1;
+    }
+    else if (a.value.length === b.value.length) {
+        if (a.index < b.index) {
+            return -1;
+        }
+        else if (a.index === b.index) {
+            return 0;
+        }
+        else if (b.index < a.index) {
+            return 1;
+        }
+    }
+    else {
+        return 1;
+    }
+};
+
+/**
+ *
+ * Break down a string into an array of "phrase segments", ordered from longest (the original string) to shortest (the
+ * shortest individual word).  Each "phrase segment" looks like:
+ * {
+ *  value: "a word or phrase, possibly including whitespace ",
+ *  index: 0 // The position of the phrase in `originalString`
+ * }
+ *
+ * @param originalString {String} - The original string to break down.
+ * @returns {Array} - An array of "phrase segments".
+ *
+ */
+gpii.diff.extractPhraseSegments = function (originalString) {
+    var regexp = /(\b[\w']+\b)(\W+)/gm;
+    var currentPhraseDef = regexp.exec(originalString);
+
+    // This was an empty string or undefined, it's not possible to match it with anything.
+    if (!originalString || originalString.length === 0) {
+        return [];
+    }
+    // There is only one "word" in this string.
+    else if (currentPhraseDef === null) {
+        return [{
+            value: originalString,
+            index: 0
+        }];
+    }
+    else {
+        var segments = [];
+        // Check for leading whitespace first.
+        var leadingWhitespaceMatches = originalString.match(/^(\W+)/);
+        if (leadingWhitespaceMatches) {
+            segments.push({
+                value: leadingWhitespaceMatches[1],
+                index: 0
+            });
+        }
+        while (currentPhraseDef !== null) {
+            // The first capture match is a "word"
+            var word = currentPhraseDef[1];
+            segments.push({
+                value: word,
+                index: currentPhraseDef.index
+            });
+            // The second (optional) match is the trailing whitespace.
+            var trailingWhitespace = currentPhraseDef[2];
+            if (trailingWhitespace.length) {
+                segments.push({
+                    value: trailingWhitespace,
+                    index: currentPhraseDef.index + word.length
+                });
+            }
+            currentPhraseDef = regexp.exec(originalString);
+        }
+        // Look for any trailing non-whitespace material after the initial pass.
+        var trailingNonWhitespaceRegexp = /\w+\W+(\w+)$/m;
+        var trailingMatches = originalString.match(trailingNonWhitespaceRegexp);
+        if (trailingMatches) {
+            segments.push({
+                value: trailingMatches[1],
+                index: originalString.length - trailingMatches[1].length
+            });
+        }
+
+        var combinedSegments = [];
+        for (var a = 0; a < segments.length; a++) {
+            for (var b = a + 1; b <= segments.length; b++) {
+                if (a + b > 0) {
```

Comment:
This is a very strange condition which looks like it is always true. Perhaps a typo or remnant of an earlier impl which had different loop bounds?

Reply:
Yes, previously it was possible for a and b to both be zero.  Removed.